examples/active_replication/flux_ingress_operator.py

```diff
@@ -0,0 +1,77 @@
+from erdos.data_stream import DataStream
+from erdos.op import Op
+from erdos.utils import setup_logging
+from flux_utils import is_ack_stream, is_control_stream, is_not_ack_stream, is_not_control_stream, is_not_back_pressure
+from flux_buffer import Buffer
+import flux_utils
+import threading
+
+
+class FluxIngressOperator(Op):
+    def __init__(self,
+                 name,
+                 output_stream_names,
+                 num_replics=2,
+                 log_file_name=None):
+        super(FluxIngressOperator, self).__init__(name)
+        self._logger = setup_logging(self.name, log_file_name)
+        self._num_replicas = num_replics
+        self._output_streams = output_stream_names
+        self._input_msg_seq_num = 0
+        self.buffer = Buffer(num_replics)   # buffer to store unacknowledged tuples
+        self._status = {}
+        for i in range(self._num_replicas):
+            self._status[i] = flux_utils.FluxOperatorState.ACTIVE
+        self.lock = threading.Lock()
+
+    @staticmethod
+    def setup_streams(input_streams, output_stream_names):
+        input_streams.filter(is_not_ack_stream)\
+            .filter(is_not_control_stream)\
+            .filter(is_not_back_pressure)\
+            .add_callback(FluxIngressOperator.on_msg)     # Input
+        input_streams.filter(is_ack_stream).add_callback(
+            FluxIngressOperator.on_ack_msg)     # Ack
+        input_streams.filter(is_control_stream).add_callback(
+            FluxIngressOperator.on_control_msg)     # Control
+
+        return [DataStream(name=output_stream_names)]
+
+    def on_msg(self, msg):
+        self.lock.acquire()
```

Comment:
Could you please explain why are you locking here? Are you assuming that on_msg can be called simultaneously by multiple threads?

Reply:
hmm my understanding is that on_msg and on_ack_msg will both change state of the op, but we want any change of state to be atomic, so this is preventing on_msg and on_ack_msg try to change state at the same time.

Reply:
I think this case cannot happen with our current ROS & Ray implementations:
1. Ray invokes the callback sequentially
2. Similarly, ROS uses a single thread to invoke them because we use rospy.spin() (see [http://wiki.ros.org/roscpp/Overview/Callbacks%20and%20Spinning](url))

Let's leave the locks in for now, but please add a comment explaining that we're looking in case in the future we'll change the ROSExecutor to use a multi threaded spinner.